### PR TITLE
Chocolatey Helper

### PR DESCRIPTION
**Pull request by elduddz**
# Description
This is looking at the issues we have been having building a successful image in UK South. #507 

This is a mix of an improvement and a change to start looking at behaviours around protecting the process from intermittent network outages, now granted this process would need to be applied all over the place, but I think this will aid the success rate of completions in places other than CUS.

The issue I see is that when an error occurs no matter how far down the line, the packer build asks you what to do or gives up, if you have gone off to do something else, 5 hours later you find the build it stuck waiting for you to abort or retry, this isnt an option for a pipeline, you only get the option of ask, which pauses indefinately or cleanup, there isnt a retry.

We are trying to automate this process, so having the build ask for what you want to do isnt really an option, so we are defaulting to destroy everything. I guess if packer had an auto retry (or at least I havent found the option)

## caveats:
I am only reviewing the win19 build as that is the one I need.

This is not a complete coverage, I have only applied fixes on the steps where the build has dropped out, I want to know if I am on the right track and there are still elements outside of this project that are pulled in.

I also noticed that packer has a little message saying vhd will be deprecated, so I have updated the json file to use managed disks, as I am putting this in a Shared image gallery, this saves a step.

#### Related issue: 
#507

## Check list
- [x] Related issue / work item is attached
- [ ] Tests are written (if applicable)
- [ ] Documentation is updated (if applicable)
- [x] Changes are tested and related VM images are successfully generated
   - this still builds in CUS, as I still cannot get a build in UKS, but this does get the process further down the line, but as the network download / outage is so intermittent, one package that might have been ok on one pass now fails.

### CODE REVIEW

**Comment by elduddz** [comment]
an alternative is a request I have raised with hashicorp:
https://github.com/hashicorp/packer/issues/8972#issuecomment-606459263

maybe we can get some likes and pleases?

File: images/win/scripts/ImageHelpers/ImageHelpers.psm1 [hunk 1]
```diff
@@ -4,6 +4,7 @@ param()
 . $PSScriptRoot\PathHelpers.ps1
 . $PSScriptRoot\InstallHelpers.ps1
 . $PSScriptRoot\MarkdownHelpers.ps1
+. $PSScriptRoot\InstallChoco.ps1
```
Comment (maxim-lobanov) [inline images/win/scripts/ImageHelpers/ImageHelpers.psm1:7]: I suggest to rename it to `ChocoHelpers.ps1` so this file can be reused for more choco-related helpers in future

File: images/win/scripts/Installers/Install-PowershellCore.ps1 [hunk 2]
```diff
@@ -2,5 +2,4 @@
 ##  File:  Install-PowershellCore.ps1
 ##  Desc:  Install PowerShell Core
 ################################################################################
-
-Invoke-Expression "& { $(Invoke-RestMethod https://aka.ms/install-powershell.ps1) } -UseMSI -Quiet"
+Install-Choco "powershell-core"
```
Comment (maxim-lobanov) [inline images/win/scripts/Installers/Install-PowershellCore.ps1:5]: What is the issue with installation from `https://aka.ms/install-powershell.ps1`. I didn't notice any issues with it previously.

File: images/win/scripts/ImageHelpers/ImageHelpers.psm1 (hunk 1 above)
Comment (elduddz) [reply images/win/scripts/ImageHelpers/ImageHelpers.psm1]: ok I have updated this file

File: images/win/scripts/ImageHelpers/MarkdownHelpers.ps1 [hunk 3]
```diff
@@ -1,5 +1,5 @@
 function Add-ContentToMarkdown {
-    [CmdletBinding()]
+
```
Comment (elduddz) [inline images/win/scripts/ImageHelpers/MarkdownHelpers.ps1:2]: Is this supposed to be here? I just noticed I am deleting the binding? but I never touched this file myself?

File: images/win/scripts/Installers/Install-PowershellCore.ps1 (hunk 2 above)
Comment (maxim-lobanov) [reply images/win/scripts/Installers/Install-PowershellCore.ps1:5]: I have looked at it deeper, it looks like both [install-powershell.ps1](https://raw.githubusercontent.com/PowerShell/PowerShell/master/tools/install-powershell.ps1) and [choco install powershell-core](https://gitlab.com/DarwinJS/ChocoPackages/-/blob/d57372d292b4166313363fa262460bccb2ad29e9/pwshpackages/powershell-core/tools/chocolateyinstall.ps1) do the same things: download installer from `https://github.com/PowerShell/PowerShell/releases/download/v${release}/${packageName}` and apply it.
The single difference: `install-powershell.ps1` determines latest version in [runtime](https://raw.githubusercontent.com/PowerShell/PowerShell/master/tools/metadata.json) but in `choco install powershell-core` the version is [hardcoded](https://gitlab.com/DarwinJS/ChocoPackages/-/commit/d57372d292b4166313363fa262460bccb2ad29e9) and package owner has to update it manually every time when the new version is released.
Considering this point, I would prefer to continue using `https://aka.ms/install-powershell.ps1`.
@alepauly for any thoughts

**Comment by maxim-lobanov** [comment]
@elduddz , just a gentle reminder :)
We would like to apply this contribution and extend this approach to more choco install invocations in future.
But there is the latest comment that prevent us to merge this pull-request (about PowerShell Core installation). Could you please revisit my comment when you have a chance?

Feel free to just revert `Install-PowershellCore.ps1` file and we will take care about retry logic there from our side.

**Comment by elduddz** [comment]
> @elduddz , just a gentle reminder :)
> We would like to apply this contribution and extend this approach to more choco install invocations in future.
> But there is the latest comment that prevent us to merge this pull-request (about PowerShell Core installation). Could you please revisit my comment when you have a chance?
> 
> Feel free to just revert `Install-PowershellCore.ps1` file and we will take care about retry logic there from our side.

Ah sorry, slipped my mind. I'll sort it out tonight.

File: images/win/scripts/Installers/Install-PowershellCore.ps1 (hunk 2 above)
Comment (elduddz) [reply images/win/scripts/Installers/Install-PowershellCore.ps1]: reverted

**Comment by maxim-lobanov** [comment]
Both checks are green. I am merging this one.
Thank you for contribution!

File: images/win/scripts/ImageHelpers/ChocoHelpers.ps1 [hunk 4]
```diff
@@ -0,0 +1,33 @@
+function Install-Choco {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory)]
+        [string]$install,
```
Comment (al-cheb) [inline images/win/scripts/ImageHelpers/ChocoHelpers.ps1:5]: nitpick: Use upper-case for parameters

Comment (al-cheb) [inline images/win/scripts/ImageHelpers/ChocoHelpers.ps1:5]: Could you please rename to be more clear? For example PackageName.

File: images/win/scripts/ImageHelpers/ChocoHelpers.ps1 [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        [int]$retries = 5
+    )
+
+    begin { }
+    process {
+        $condition = $false
+        $count = 0
+        do {
+            Write-Output "running: powershell choco install $install -y"
+            powershell choco install $install -y
```
Comment (al-cheb) [inline images/win/scripts/ImageHelpers/ChocoHelpers.ps1:15]: Why do we need to run another instance of powershell?

**Pull request merged**